File: packages/app/src/services/software/definitions/variants/plexamp.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'plexamp';
	public name = 'Plexamp';
	public category = [SoftwareCategory.Audio, SoftwareCategory.Media];
	public downloadName = 'Plexamp Setup.exe';
	public shouldCacheUrl = true;
	public icon = 'plexamp.png';
	public homepage = 'https://plex.tv';
	public parent = 'plex';

	public async resolveDownloadUrl() {
		const res = await fetch('https://www.plex.tv/wp-json/plex/v1/downloads/plexamp');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/Plexamp%20Setup%20\d+\.\d+\.\d+\.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(`https://plexamp.plex.tv/plexamp.plex.tv/desktop/${match[0]}`);
	}
}

File: src/lib/software/variants/plexamp.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class Plexamp(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'plexamp'
        self.name = 'Plexamp'
        self.download_name = 'Plexamp Setup.exe'
        self.should_cache_url = True
        self.icon = 'plexamp.png'
        self.homepage = 'https://plex.tv'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'https:\\/\\/plexamp\.plex\.tv\\/plexamp\.plex\.tv\\/desktop\\/Plexamp%20Setup%20\d+\.\d+\.\d+\.exe')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(0).replace('\\', ''))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://www.plex.tv/wp-json/plex/v1/downloads/plexamp'))
